### PR TITLE
Feature/api key authentication

### PR DESCRIPTION
**Pull request by bretterer**
ApiKey Authentication and OAuth tokens added

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -8,11 +8,6 @@ php:
 
 sudo: false
 
-env:
```
Comment (ecrisostomo) [inline .travis.yml]: Why was this removed from here? What are we using to run the tests in travis?

File: README.md [hunk 2]
```diff
@@ -1277,6 +1277,171 @@ $providerData->modifiedAt; //-> 2014-04-01T17:00:09.189Z
...
+
+  grant_type=client_credentials
+```
+
+The `Authorization` header contains a base64 encoding of the API Key and Secret.
+
+```php
+$application = \Stormpath\Resource\Application::get("https://api.stormpath.com/v1/applications/24mp4us71ntza6lBwlu");
+
+$request = \Stormpath\Authc\Api\Request::createFromGlobals();
```
Comment (ecrisostomo) [inline README.md]: Is there any other way for the user to create the request that is more flexible?

File: src/Stormpath/Authc/Api/ApiAuthenticationResult.php [hunk 3]
```diff
@@ -0,0 +1,9 @@
+<?php
+
+namespace Stormpath\Authc\Api;
+
+
+class ApiAuthenticationResult extends AuthenticatorResult
+{
+
+}
```
Comment (ecrisostomo) [inline src/Stormpath/Authc/Api/ApiAuthenticationResult.php]: Any reasons to have this empty class?

File: src/Stormpath/Authc/Api/BasicAuthenticationResult.php [hunk 4]
```diff
@@ -0,0 +1,9 @@
+<?php
+
+namespace Stormpath\Authc\Api;
+
+
+class BasicAuthenticationResult extends AuthenticatorResult
+{
+
+}
```
Comment (ecrisostomo) [inline src/Stormpath/Authc/Api/BasicAuthenticationResult.php]: Any reason to have this empty class?

File: src/Stormpath/Resource/Account.php [hunk 5]
```diff
@@ -207,4 +208,22 @@ public function delete() {
...
+        $apiKey = $this->getDataStore()->create($this->getHref() . '/' . ApiKey::PATH,
+            $apiKey, Stormpath::API_KEY, $options);
+
+        if ($apiKey)
+        {
+            $apiKey->setApiKeyMetadata($apiKeyOptions);
+        }
+
+        return $apiKey;
+    }
```
Comment (ecrisostomo) [inline src/Stormpath/Resource/Account.php]: Looks like the api key collection was not added to the account resource.

File: tests/Stormpath/Tests/Resource/ApplicationTest.php [hunk 6]
```diff
@@ -526,6 +526,42 @@ public function testSave()
...
+        $apiKey->status = 'DISABLED';
+        $apiKey->save();
+        $this->assertEquals('DISABLED', $apiKey->status);
+
+        $apiKey->delete();
+        $apiKey = $application->getApiKey($newApiKey->id);
+        $this->assertNull($apiKey);
+
+        $account->delete();
+    }
```
Comment (ecrisostomo) [inline tests/Stormpath/Tests/Resource/ApplicationTest.php]: We need to add tests to verify that when an api key is retrieved from the application, it is actually retrieved from the cache. This is something that doesn't look to be implemented at all because this requires that the data store caches the api key when created, when retrieved from the account api key collection (which I think needs to be added) and when retrieved by api key id from the application (this requires a special implementation for cacheing). This last scenario (retrieving the api key from the application) is specially important since for authentication requests we always need to perform this query and we shouldn't need to call the server every time because the api keys never changes, except for when the status changes or it gets deleted at which point the cache should get updated.

File: src/Stormpath/Authc/Api/OAuthAuthenticationResult.php [hunk 7]
```diff
@@ -0,0 +1,9 @@
+<?php
+
+namespace Stormpath\Authc\Api;
+
+
+class OAuthAuthenticationResult extends AuthenticatorResult
+{
+
+}
```
Comment (ecrisostomo) [inline src/Stormpath/Authc/Api/OAuthAuthenticationResult.php]: Any reasons to have this empty class?

File: src/Stormpath/Authc/Api/OAuthBearerAuthenticationResult.php [hunk 8]
```diff
@@ -0,0 +1,9 @@
+<?php
+
+namespace Stormpath\Authc\Api;
+
+
+class OAuthBearerAuthenticationResult extends AuthenticatorResult
+{
+
+}
```
Comment (ecrisostomo) [inline src/Stormpath/Authc/Api/OAuthBearerAuthenticationResult.php]: Any reasons to have this empty class?

File: src/Stormpath/Resource/ApiKey.php [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+ * limitations under the License.
+ */
+
+namespace Stormpath\Resource;
+
+
+use Stormpath\Authc\Api\ApiKeyEncryptionUtils;
+use Stormpath\Stormpath;
+
+class ApiKey extends InstanceResource implements Deletable
```
Comment (ecrisostomo) [inline src/Stormpath/Resource/ApiKey.php]: In favor of this api key resource, we should be deprecating the one on the Stormpath folder and document that it will go away after the first final version (when we remove "beta"). Something similar to what we did on the Java SDK. This is because these two classes represent api keys, only that one is a Resource instance and the other one is not. 

That means that, after removing the deprecated class, this class will also be the one used for authenticating the SDK calls to Stormpath.

File: tests/Stormpath/Tests/Resource/ApplicationTest.php (hunk 6 above)
Comment (ecrisostomo) [inline tests/Stormpath/Tests/Resource/ApplicationTest.php]: Should also make sure that the api key secret is always encrypted in the cache.

File: .travis.yml (hunk 1 above)
Comment (bretterer) [inline .travis.yml]: I moved the keys into the travis config for easier changes if needed.

File: src/Stormpath/Authc/Api/ApiAuthenticationResult.php (hunk 3 above)
Comment (bretterer) [inline src/Stormpath/Authc/Api/ApiAuthenticationResult.php]: The class is just a wrapper of AuthenticatorResult so we get the correct return type per the spec.

File: src/Stormpath/Authc/Api/BasicAuthenticationResult.php (hunk 4 above)
Comment (bretterer) [inline src/Stormpath/Authc/Api/BasicAuthenticationResult.php]: The class is just a wrapper of AuthenticatorResult so we get the correct return type per the spec.

File: src/Stormpath/Resource/Account.php (hunk 5 above)
Comment (bretterer) [inline src/Stormpath/Resource/Account.php]: This was all part of stuff that was added by Diego

File: src/Stormpath/Authc/Api/OAuthAuthenticationResult.php (hunk 7 above)
Comment (bretterer) [inline src/Stormpath/Authc/Api/OAuthAuthenticationResult.php]: The class is just a wrapper of AuthenticatorResult so we get the correct return type per the spec.

File: src/Stormpath/Authc/Api/OAuthBearerAuthenticationResult.php (hunk 8 above)
Comment (bretterer) [inline src/Stormpath/Authc/Api/OAuthBearerAuthenticationResult.php]: The class is just a wrapper of AuthenticatorResult so we get the correct return type per the spec.

File: .travis.yml (hunk 1 above)
Comment (ecrisostomo) [inline .travis.yml]: There hasn't been any need to change that so far, so I think is still ok if we leave it here. Also, it's under version control here so we know when it changes and we know it's not going to get lost (thinking that travis might lose it for whatever crazy reason). So, as part of the changes for this implementation, please revert this change to add the keys back to this file.

File: src/Stormpath/Authc/Api/ApiAuthenticationResult.php (hunk 3 above)
Comment (ecrisostomo) [inline src/Stormpath/Authc/Api/ApiAuthenticationResult.php]: On the spec I only see that a "result" is expected, not necessarily  a different result type per authenticator type. I think as a general good practice, if a class has no implementation at all, it's very likely that it is not needed; in the case of this implementation there are multiple classes that are unnecessary (all the ones that extend `AuthenticatorResult`).

I think we still have time to remove these classes because per the spec on the README, people are just expecting a "result" which could still be an `AuthenticatorResult`.

File: src/Stormpath/Resource/Account.php (hunk 5 above)
Comment (ecrisostomo) [inline src/Stormpath/Resource/Account.php]: Alright, but we still need to add the api key collection to the account resource.

File: .travis.yml (hunk 1 above)
Comment (bretterer) [inline .travis.yml]: No problem, I did this in regards to having to change the keys in general anyways per @omgitstom 's request to all of us to keep testing within the same tenant

File: src/Stormpath/Authc/Api/ApiAuthenticationResult.php (hunk 3 above)
Comment (bretterer) [inline src/Stormpath/Authc/Api/ApiAuthenticationResult.php]: @ecrisostomo actually, the spec shows different output for each type of call.  Check out the spec @omgitstom is showing here https://github.com/stormpath/stormpath-sdk-spec/blob/master/specifications/api_auth.md

Comment (ecrisostomo) [inline src/Stormpath/Authc/Api/ApiAuthenticationResult.php]: Interesting. I was talking about the README specs, but I get what you are saying now. However, if those classes  don't have any implementation I don't see a reason to have them. Having said that, I think we could give them some use after taking a look at the Java SDK where each of the subclasses have some difference (usually getting a property that is unique for that type of result). See the interfaces on the Java SDK and their differences:

* [AuthenticationResult](https://github.com/stormpath/stormpath-sdk-java/blob/master/api/src/main/java/com/stormpath/sdk/authc/AuthenticationResult.java)
* [ApiAuthenticationResult](https://github.com/stormpath/stormpath-sdk-java/blob/master/api/src/main/java/com/stormpath/sdk/api/ApiAuthenticationResult.java)
* [AccessTokenResult](https://github.com/stormpath/stormpath-sdk-java/blob/master/api/src/main/java/com/stormpath/sdk/oauth/AccessTokenResult.java)
* [OauthAuthenticationResult](https://github.com/stormpath/stormpath-sdk-java/blob/master/api/src/main/java/com/stormpath/sdk/oauth/OauthAuthenticationResult.java)

So, we should do something similar or just use the one `AuthenticatorResult` that we have (which should have been called `AuthenticationResult`, but we'll roll with it).
